core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -319,21 +341,49 @@ public Response postBucket(@PathParam("bucket") final String bucket,
   }
 
   /**
-   * @summary creates a bucket
+   * Creates a bucket, or puts bucket tags on an existing bucket.
    * @param authorization header parameter authorization
    * @param bucket the bucket name
+   * @param tagging query string to indicate if this is for PutObjectTagging or not
+   * @param is the request body
    * @return the response object
    */
   @PUT
   @Path(BUCKET_PARAM)
+  @Consumes(MediaType.APPLICATION_XML)
   public Response createBucket(@HeaderParam("Authorization") String authorization,
-                               @PathParam("bucket") final String bucket) {
+                               @PathParam("bucket") final String bucket,
+                               @QueryParam("tagging") final String tagging,
+                               final InputStream is) {
     return S3RestUtils.call(bucket, () -> {
       Preconditions.checkNotNull(bucket, "required 'bucket' parameter is missing");
+      final FileSystem fs = getFileSystem(authorization);
+      String bucketPath = S3RestUtils.parsePath(String.format("%s%s",
+          AlluxioURI.SEPARATOR, bucket));
 
-      String bucketPath = S3RestUtils.parsePath(AlluxioURI.SEPARATOR + bucket);
+      if (tagging != null) { // PutBucketTagging
+        S3RestUtils.checkPathIsAlluxioDirectory(fs, bucketPath);
+        try {
+          TaggingData tagData = new XmlMapper().readerFor(TaggingData.class)
+              .readValue(is);
+          LOG.debug("PutBucketTagging tagData={}", tagData);
+          Map<String, ByteString> xattrMap = new HashMap<String, ByteString>();
+          xattrMap.put(S3Constants.TAGGING_XATTR_KEY, TaggingData.serialize(tagData));
+          SetAttributePOptions attrPOptions = SetAttributePOptions.newBuilder()
+              .putAllXattr(xattrMap).setXattrUpdateStrategy(File.XAttrUpdateStrategy.UNION_REPLACE)
+              .build();
+          fs.setAttribute(new AlluxioURI(bucketPath), attrPOptions);
+        } catch (IOException e) {
+          if (e.getCause() instanceof S3Exception) {
+            throw S3RestUtils.toBucketS3Exception((S3Exception) e.getCause(), bucketPath);
+          }
+          throw new S3Exception(e, bucketPath, S3ErrorCode.MALFORMED_XML);
+        } catch (Exception e) {
+          throw S3RestUtils.toBucketS3Exception(e, bucketPath);
+        }
+        return Response.Status.OK;
+      }
 
-      final FileSystem fs = getFileSystem(authorization);
       // Create the bucket.
       CreateDirectoryPOptions options =
           CreateDirectoryPOptions.newBuilder().setWriteType(S3RestUtils.getS3WriteType()).build();
```

Comment:
It depends on what we prefer, the code structure for these methods is very messy because multiple S3 API Actions are matching the same JAX-RS handler.

I decided I would prefer to isolate entire requests into their own control-flow block where possible as opposed to mixing implementations. Hence I have the `if (tagging != null)` block return on its own.